pkg/cloud/services/ec2/launchtemplate_test.go

```diff
@@ -100,45 +121,151 @@ func TestGetLaunchTemplate(t *testing.T) {
+				if err != nil {
+					t.Fatalf("did not expect error: %v", err)
+				}
+				if userDataHash != "" {
+					t.Fatalf("Did not expect a userdata hash, but got something: %s", userDataHash)
+				}
+				if launchTemplate != nil {
+					t.Fatalf("Did not expect anything but got something: %+v", launchTemplate)
+				}
+			},
+		},
+		{
+			name:               "successfully fetched launch template versions",
+			launchTemplateName: "foo",
+			expect: func(scope *mocks.MockEC2ScopeMockRecorder, m *mock_ec2iface.MockEC2APIMockRecorder) {
+				scope.V(2).Return(klogr.New())
+				m.DescribeLaunchTemplateVersions(gomock.Eq(&ec2.DescribeLaunchTemplateVersionsInput{
+					LaunchTemplateName: aws.String("foo"),
+					Versions:           []*string{aws.String("$Latest")},
+				})).
+					Return(&ec2.DescribeLaunchTemplateVersionsOutput{
+						LaunchTemplateVersions: []*ec2.LaunchTemplateVersion{
+							{
+								LaunchTemplateId:   aws.String("lt-12345"),
+								LaunchTemplateName: aws.String("foo"),
+								LaunchTemplateData: &ec2.ResponseLaunchTemplateData{
+									ImageId: aws.String("foo-image"),
+									IamInstanceProfile: &ec2.LaunchTemplateIamInstanceProfileSpecification{
+										Arn: aws.String("instance-profile/foo-profile"),
+									},
+									KeyName: aws.String("foo-keyname"),
+									BlockDeviceMappings: []*ec2.LaunchTemplateBlockDeviceMapping{
+										{
+											DeviceName: aws.String("foo-device"),
+											Ebs: &ec2.LaunchTemplateEbsBlockDevice{
+												Encrypted:  aws.Bool(true),
+												VolumeSize: aws.Int64(16),
+												VolumeType: aws.String("cool"),
+											},
+										},
+									},
+									NetworkInterfaces: []*ec2.LaunchTemplateInstanceNetworkInterfaceSpecification{
+										{
+											DeviceIndex: aws.Int64(1),
+											Groups:      []*string{aws.String("foo-group")},
+										},
+									},
+									UserData: aws.String(base64.StdEncoding.EncodeToString([]byte(testUserData))),
+								},
+								VersionNumber: aws.Int64(1),
+							},
+						},
+					}, nil)
+			},
+			check: func(launchTemplate *expinfrav1.AWSLaunchTemplate, userDataHash string, err error) {
+				wantLT := &expinfrav1.AWSLaunchTemplate{
+					Name: "foo",
+					AMI: infrav1.AMIReference{
+						ID: aws.String("foo-image"),
+					},
+					IamInstanceProfile: "foo-profile",
+					SSHKeyName:         aws.String("foo-keyname"),
+					VersionNumber:      aws.Int64(1),
 				}
 
-				if launchtemplate != nil {
-					t.Fatalf("Did not expect anything but got something: %+v", launchtemplate)
+				if !reflect.DeepEqual(launchTemplate, wantLT) {
+					t.Fatalf("launchTemplate mismatch: got %v, want %v", launchTemplate, wantLT)
+				}
+				if !reflect.DeepEqual(userDataHash, testUserDataHash) {
+					t.Fatalf("userDataHash mismatch: got %v, want %v", userDataHash, testUserDataHash)
+				}
+				if err != nil {
+					t.Fatalf("did not expect error: %v", err)
 				}
 			},
 		},
```

Comment:
Can we also add a scenario:
Should return computed userdata if userdata is nil in launchtemplate